README.md
# PHP dotenv-vault [![PHP version](https://badge.fury.io/ph/dotenv-org%2Fphpdotenv-vault.svg)](https://badge.fury.io/ph/dotenv-org%2Fphpdotenv-vault)

<img src="https://raw.githubusercontent.com/motdotla/dotenv/master/dotenv.svg" alt="dotenv-vault" align="right" width="200" />

Extends the proven & trusted foundation of [phpdotenv](https://github.com/vlucas/phpdotenv), with a `.env.vault` file.

The extended standard lets you load encrypted secrets from your `.env.vault` file in production (and other) environments. Brought to you by the same people that pioneered [dotenv-nodejs](https://github.com/motdotla/dotenv).

* [🌱 Install](#-install)
* [🏗️ Usage (.env)](#%EF%B8%8F-usage)
* [🚀 Deploying (.env.vault) 🆕](#-deploying)
* [🌴 Multiple Environments](#-manage-multiple-environments)
* [❓ FAQ](#-faq)
* [⏱️ Changelog](./CHANGELOG.md)

## 🌱 Install

```shell
$ composer require dotenv-org/phpdotenv-vault
```

## 🏗️ Usage

Development usage works just like [phpdotenv](https://github.com/vlucas/phpdotenv).

Add your application configuration to your `.env` file in the root of your project:

```shell
# .env
S3_BUCKET="dotenv"
SECRET_KEY="souper_seekret_key"
```

As early as possible in your application bootstrap process, load .env:

```php
require 'vendor/autoload.php';

$dotenv = DotenvVault\DotenvVault::createImmutable([__DIR__]);
$dotenv->safeLoad();
```

When your application loads, these variables will be available in `$_SERVER`:

```php
$s3_bucket = $_SERVER['S3_BUCKET'];
echo $s3_bucket;
```

## 🚀 Deploying

Encrypt your environment variables by doing:

```shell
npx dotenv-vault local build
```

This will create an encrypted `.env.vault` file along with a `.env.keys` file containing the encryption keys. Set the `DOTENV_KEY` environment variable by copying and pasting the key value from the `.env.keys` file onto your server or cloud provider. For example in heroku:

```shell
heroku config:set DOTENV_KEY=<key string from .env.keys>
```

Commit your .env.vault file safely to code and deploy. Your .env.vault fill be decrypted on boot, its environment variables injected, and your app work as expected.

Note that when the `DOTENV_KEY` environment variable is set, environment settings will *always* be loaded from the `.env.vault` file in the project root. For development use, you can leave the `DOTENV_KEY` environment variable unset and fall back on the `dotenv` behaviour of loading from `.env`.

## 🌴 Manage Multiple Environments

You have two options for managing multiple environments - locally managed or vault managed - both use [dotenv-vault](https://github.com/dotenv-org/dotenv-vault).

Locally managed never makes a remote API call. It is completely managed on your machine. Vault managed adds conveniences like backing up your .env file, secure sharing across your team, access permissions, and version history. Choose what works best for you.

#### 💻 Locally Managed

Create a `.env.production` file in the root of your project and put your production values there.

```shell
# .env.production
S3_BUCKET="PRODUCTION_S3BUCKET"
SECRET_KEY="PRODUCTION_SECRETKEYGOESHERE"
```

Rebuild your `.env.vault` file.

```shell
npx dotenv-vault local build
```

View your `.env.keys` file. There is a production `DOTENV_KEY` that pairs with the `DOTENV_VAULT_PRODUCTION` cipher in your `.env.vault` file.

Set the production `DOTENV_KEY` on your server, recommit your `.env.vault` file to code, and deploy. That's it!

Your .env.vault fill be decrypted on boot, its production environment variables injected, and your app work as expected.

#### 🔐 Vault Managed

Sync your .env file. Run the push command and follow the instructions. [learn more](/docs/sync/quickstart)

```
$ npx dotenv-vault push
```

Manage multiple environments with the included UI. [learn more](/docs/tutorials/environments)

```
$ npx dotenv-vault open
```

Build your `.env.vault` file with multiple environments.

```
$ npx dotenv-vault build
```

Access your `DOTENV_KEY`.

```
$ npx dotenv-vault keys
```

Set the production `DOTENV_KEY` on your server, recommit your `.env.vault` file to code, and deploy. That's it!

## ❓ FAQ

#### What happens if `DOTENV_KEY` is not set?

Dotenv Vault gracefully falls back to [phpdotenv](https://github.com/vlucas/phpdotenv) when `DOTENV_KEY` is not set. This is the default for development so that you can focus on editing your `.env` file and save the `build` command until you are ready to deploy those environment variables changes.

#### Should I commit my `.env` file?

No. We **strongly** recommend against committing your `.env` file to version control. It should only include environment-specific values such as database passwords or API keys. Your production database should have a different password than your development database.

#### Should I commit my `.env.vault` file?

Yes. It is safe and recommended to do so. It contains your encrypted envs, and your vault identifier.

#### Can I share the `DOTENV_KEY`?

No. It is the key that unlocks your encrypted environment variables. Be very careful who you share this key with. Do not let it leak.

## Contributing

1. Fork it
2. Create your feature branch (`git checkout -b my-new-feature`)
3. Commit your changes (`git commit -am 'Added some feature'`)
4. Push to the branch (`git push origin my-new-feature`)
5. Create new Pull Request

## Changelog

See [CHANGELOG.md](CHANGELOG.md)

## License

BSD-3


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file. See [standard-version](https://github.com/conventional-changelog/standard-version) for commit guidelines.

## [Unreleased](https://github.com/dotenv-org/phpdotenv-vault/compare/v0.2.4...master)

## 0.2.4

### Changed

- Move logging to `error_log` rather than `echo`.

## 0.2.3

### Added

- echo/log out the dotenv-vault decryption message

## 0.2.2

### Added

- Added additional createX methods
- Added tests

## 0.2.1

### Changed

- Added support for passing string to paths argument.

## 0.2.0

### Added

- Moved decryption to its own class for better testing and ease of usage

### Fixed

- DOTENV_KEY was not respected if set in the infrastructure. Fixed.
- Decryptiong could fail related to some misconfigured logic. Fixed.

## 0.1.3

### Removed

- Remove `var_dump` when falling back to `.env` file

## 0.1.2

### Changed

- Use `getenv` instead of `$_ENV`

## 0.1.1

### Removed

- Remove laravel service provider that is not hooked up

## 0.1.0

### Added

- Basic functionality for decrypting `.env.vault` files supported


tests/DotEnvVaultTest.php
<?php

namespace DotenvVault\Tests;

use Dotenv\Loader\LoaderInterface;
use Dotenv\Parser\Entry;
use Dotenv\Parser\ParserInterface;
use Dotenv\Parser\Value;
use Dotenv\Repository\RepositoryInterface;
use Dotenv\Store\StoreInterface;
use DotenvVault\DotEnvVault;
use DotenvVault\DotEnvVaultError;
use Mockery;
use Mockery\Adapter\Phpunit\MockeryTestCase;

class DotEnvVaultTest extends MockeryTestCase
{
    private $vaultKeyDev = 'DOTENV_VAULT_DEVELOPMENT';
    private $keyDev = 'dotenv://:key_12428c2d50a7bf9f0eef6cf4c26dd6ac6b88bf5bc081dd859f6e5a0ab3a56aae@dotenv.local/vault/.env.vault?environment=development';
    private $encryptedDev = 'p8xMNWI8rdR8pMHdsUdD70fIKnpIws/bx2ur+LrP/QEljrCN9EowXaH5+4t2jg6EuSljOb+TLQ==';
    private $decryptedDevStr = <<<ENV
APP_ENV=development
FOO=bar
ENV;
    private $decryptedDevArr = ['APP_ENV' => 'development', 'FOO' => 'bar'];

    private $vaultKeyProd = 'DOTENV_VAULT_PRODUCTION';
    private $keyProd = 'dotenv://:key_f99471b404c2cb6e321f7fd97abba8be17a343b429cef02e615db8576f11d310@dotenv.local/vault/.env.vault?environment=production';
    private $encryptedProd = '0EWcDIA9YidGiRtMMag/ggftCEfA+hKG5gGYC5Z5bXue/20uM850AgRuMhhbhfGkvZatz+yX';
    private $decryptedProdStr = <<<ENV
FOO=baz
APP_ENV=production
ENV;
    private $decryptedProdArr = ['FOO' => 'baz', 'APP_ENV' => 'production'];

    private $vaultFile;
    /** @var Entry[] */
    private $vaultFileEntries;
    /** @var string[] */
    private $vaultFileArr;

    /** @return Entry[] */
    private function makeDecryptedDevEntries(): array
    {
        return [
            new Entry('APP_ENV', Value::blank()->append('development', true)),
            new Entry('FOO', Value::blank()->append('bar', true)),
        ];
    }

    /** @return Entry[] */
    private function makeDecryptedProdEntries(): array
    {
        return [
            new Entry('FOO', Value::blank()->append('baz', true)),
            new Entry('APP_ENV', Value::blank()->append('production', true)),
        ];
    }

    protected function setUp(): void
    {
        parent::setUp();

        $this->vaultFile = <<<ENV
# .env.vault (generated with npx dotenv-vault local build)
{$this->vaultKeyDev}="{$this->encryptedDev}"
{$this->vaultKeyProd}="{$this->encryptedProd}"

ENV;
        $this->vaultFileEntries = [
            new Entry($this->vaultKeyDev, Value::blank()->append($this->encryptedDev, true)),
            new Entry($this->vaultKeyDev, Value::blank()->append($this->encryptedProd, true)),
        ];
        $this->vaultFileArr = [$this->vaultKeyDev => $this->encryptedDev, $this->vaultKeyProd => $this->encryptedProd];
    }

    /** @dataProvider getLoadWithKeyData */
    public function testLoadWithKey(
        string $key,
        string $vaultName,
        string $encryptedVault,
        string $decryptedVault,
        array $decryptedEntries,
        array $decryptedArr
    ): void {
        $repository = Mockery::mock(RepositoryInterface::class);
        $repository->allows()->get('DOTENV_KEY')->andReturn($key);
        $repository->allows()->get($vaultName)->andReturn($encryptedVault);

        $store = Mockery::mock(StoreInterface::class);
        $store->allows()->read()->andReturn($this->vaultFile)->atLeast()->once();

        $parser = Mockery::mock(ParserInterface::class);
        $parser->allows()->parse($this->vaultFile)->andReturn($this->vaultFileEntries)->atLeast()->once();
        $parser->allows()->parse($decryptedVault)->andReturn($decryptedEntries)->atLeast()->once();

        $loader = Mockery::mock(LoaderInterface::class);
        $loader->allows()->load($repository, $this->vaultFileEntries)->andReturn($this->vaultFileArr)->atLeast()->once();
        $loader->allows()->load($repository, $decryptedEntries)->andReturn($decryptedArr)->atLeast()->once();

        $inst = new DotEnvVault($store, $parser, $loader, $repository);
        $actual = $inst->load();

        $this->assertSame($decryptedArr, $actual);
    }

    public function getLoadWithKeyData(): array
    {
        $decryptedDevEntries = $this->makeDecryptedDevEntries();
        $decryptedProdEntries = $this->makeDecryptedProdEntries();

        return [
            'dev' => [
                'key' => $this->keyDev,
                'vaultName' => $this->vaultKeyDev,
                'encryptedVault' => $this->encryptedDev,
                'decryptedVault' => $this->decryptedDevStr,
                'decryptedEntries' => $decryptedDevEntries,
                'decryptedArr' => $this->decryptedDevArr,
            ],
            'prod' => [
                'key' => $this->keyProd,
                'vaultName' => 'DOTENV_VAULT_PRODUCTION',
                'encryptedVault' => $this->encryptedProd,
                'decryptedVault' => $this->decryptedProdStr,
                'decryptedEntries' => $decryptedProdEntries,
                'decryptedArr' => $this->decryptedProdArr,
            ],
            'multi keys, valid first' => [
                'key' => "{$this->keyDev},dotenv://:key_0000000000000000000000000000000000000000000000000000000000000000@dotenv.local/vault/.env.vault?environment=development",
                'vaultName' => $this->vaultKeyDev,
                'encryptedVault' => $this->encryptedDev,
                'decryptedVault' => $this->decryptedDevStr,
                'decryptedEntries' => $decryptedDevEntries,
                'decryptedArr' => $this->decryptedDevArr,
            ],
            'multi keys, valid last' => [
                'key' => "dotenv://:key_0000000000000000000000000000000000000000000000000000000000000000@dotenv.local/vault/.env.vault?environment=production,{$this->keyProd}",
                'vaultName' => $this->vaultKeyProd,
                'encryptedVault' => $this->encryptedProd,
                'decryptedVault' => $this->decryptedProdStr,
                'decryptedEntries' => $decryptedProdEntries,
                'decryptedArr' => $this->decryptedProdArr,
            ],
        ];
    }

    public function testLoadWithoutKey(): void
    {
        $repository = Mockery::mock(RepositoryInterface::class);
        $repository->allows()->get('DOTENV_KEY')->andReturn('');

        $store = Mockery::mock(StoreInterface::class);
        $store->allows()->read()->andReturn($this->vaultFile)->atLeast()->once();

        $parser = Mockery::mock(ParserInterface::class);
        $parser->allows()->parse($this->vaultFile)->andReturn($this->vaultFileEntries)->atLeast()->once();

        $loader = Mockery::mock(LoaderInterface::class);
        $loader->allows()->load($repository, $this->vaultFileEntries)->andReturn($this->vaultFileArr)->atLeast()->once();

        $inst = new DotEnvVault($store, $parser, $loader, $repository);
        $actual = $inst->load();

        $this->assertSame($this->vaultFileArr, $actual);
    }

    public function testLoadWithInvalidKey(): void
    {
        $repository = Mockery::mock(RepositoryInterface::class);
        $repository->allows()->get('DOTENV_KEY')->andReturn('dotenv://:key_0000000000000000000000000000000000000000000000000000000000000000@dotenv.local/vault/.env.vault?environment=production');
        $repository->allows()->get($this->vaultKeyProd)->andReturn($this->encryptedProd);

        $store = Mockery::mock(StoreInterface::class);
        $store->allows()->read()->andReturn($this->vaultFile)->atLeast()->once();

        $parser = Mockery::mock(ParserInterface::class);
        $parser->allows()->parse($this->vaultFile)->andReturn($this->vaultFileEntries)->atLeast()->once();

        $loader = Mockery::mock(LoaderInterface::class);
        $loader->allows()->load($repository, $this->vaultFileEntries)->andReturn($this->vaultFileArr)->atLeast()->once();

        $inst = new DotEnvVault($store, $parser, $loader, $repository);

        $this->expectExceptionObject(new DotEnvVaultError('INVALID_DOTENV_KEY: Key must be valid.'));
        $inst->load();
    }

    public function testCreate(): void
    {
        $actual = DotEnvVault::create(Mockery::mock(RepositoryInterface::class),'path', '.env.vault');

        $this->assertInstanceOf(DotEnvVault::class, $actual);
    }

    public function testCreateImmutable(): void
    {
        $actual = DotEnvVault::createImmutable('path', '.env.vault');

        $this->assertInstanceOf(DotEnvVault::class, $actual);
    }
}


tests/DotenvVault/DotenvVaultTest.php
<?php

declare(strict_types=1);

namespace DotenvVault\Tests;

// phpdotenv-vault libs
use DotenvVault\DotenvVault;

// phpdotenv libs
use Dotenv;

use PHPUnit\Framework\TestCase;

final class DotenvVaultTest extends TestCase
{
    /**
     * @var string
     */
    private static $folder;

    /**
     * @beforeClass
     *
     * @return void
     */
    public static function setFolder()
    {
        self::$folder = \dirname(__DIR__).'/fixtures/env';
    }

    protected function setUp(): void {
    }

    protected function tearDown(): void {
    }

    public function testThrowsExceptionIfUnableToLoadFile()
    {
        $dotenvVault = DotenvVault::createMutable(__DIR__);

        $this->expectExceptionMessage('Unable to read any of the environment file(s) at');

        $dotenvVault->load();
    }

    public function testTriesPathsToLoad()
    {
        $dotenvVault = DotenvVault::createImmutable([__DIR__, self::$folder]);
        self::assertCount(4, $dotenvVault->load());
    }

    public function testLoadFromEnvFile()
    {
        $dotenvVault = DotenvVault::createImmutable([__DIR__, self::$folder]);
        $dotenvVault->load();

        self::assertEquals($_ENV['FOO'], 'bar');
        self::assertEquals($_ENV['BAR'], 'baz');
    }

    public function testLoadFromEnvFileMutable()
    {
        $dotenvVault = DotenvVault::createMutable([__DIR__, self::$folder]);
        $dotenvVault->load();

        self::assertEquals($_ENV['FOO'], 'bar');
        self::assertEquals($_ENV['BAR'], 'baz');
    }

    public function testLoadFromEnvFileUnsafeMutable()
    {
        $dotenvVault = DotenvVault::createUnsafeMutable([__DIR__, self::$folder]);
        $dotenvVault->load();

        self::assertEquals(getenv('FOO'), 'bar');
        self::assertEquals(getenv('BAR'), 'baz');
    }

    public function testLoadFromEnvFileUnsafeImmutable()
    {
        $dotenvVault = DotenvVault::createUnsafeImmutable([__DIR__, self::$folder]);
        $dotenvVault->load();

        self::assertEquals(getenv('FOO'), 'bar');
        self::assertEquals(getenv('BAR'), 'baz');
    }

    public function testSafeLoadFromEnvFile()
    {
        $dotenvVault = DotenvVault::createImmutable([__DIR__, self::$folder]);
        $dotenvVault->safeLoad();

        self::assertEquals($_ENV['FOO'], 'bar');
        self::assertEquals($_ENV['BAR'], 'baz');
    }

    public function testLoadFromEnvFileAndPathsPassedAsString()
    {
        $dotenvVault = DotenvVault::createImmutable(self::$folder);
        $dotenvVault->load();

        self::assertEquals($_ENV['FOO'], 'bar');
        self::assertEquals($_ENV['BAR'], 'baz');
    }

    public function testLoadFromEnvVaultFileWhenDotenvKeyPresent()
    {
        $_ENV["DOTENV_KEY"] = 'dotenv://:key_ddcaa26504cd70a6fef9801901c3981538563a1767c297cb8416e8a38c62fe00@dotenv.org/vault/.env.vault?environment=development';

        $dotenvVault = DotenvVault::createImmutable([__DIR__, self::$folder]);
        $dotenvVault->load();

        self::assertEquals($_ENV['ALPHA'], 'zeta');
    }

    public function testLoadFromEnvVaultFileWhenDotenvKeyPresentAndPathsPassedAsString()
    {
        $_ENV["DOTENV_KEY"] = 'dotenv://:key_ddcaa26504cd70a6fef9801901c3981538563a1767c297cb8416e8a38c62fe00@dotenv.org/vault/.env.vault?environment=development';

        $dotenvVault = DotenvVault::createImmutable(self::$folder);
        $dotenvVault->load();

        self::assertEquals($_ENV['ALPHA'], 'zeta');
    }

    public function testLoadFromEnvVaultFileWhenDotenvKeyPresentMutable()
    {
        $_ENV["DOTENV_KEY"] = 'dotenv://:key_ddcaa26504cd70a6fef9801901c3981538563a1767c297cb8416e8a38c62fe00@dotenv.org/vault/.env.vault?environment=development';

        $dotenvVault = DotenvVault::createMutable([__DIR__, self::$folder]);
        $dotenvVault->load();

        self::assertEquals($_ENV['ALPHA'], 'zeta');
    }

    public function testLoadFromEnvVaultFileWhenDotenvKeyPresentUnsafeImmutable()
    {
        $_ENV["DOTENV_KEY"] = 'dotenv://:key_ddcaa26504cd70a6fef9801901c3981538563a1767c297cb8416e8a38c62fe00@dotenv.org/vault/.env.vault?environment=development';

        $dotenvVault = DotenvVault::createUnsafeImmutable([__DIR__, self::$folder]);
        $dotenvVault->load();

        self::assertEquals($_ENV['ALPHA'], 'zeta');
    }

    public function testLoadFromEnvVaultFileWhenDotenvKeyPresentUnsafeMutable()
    {
        $_ENV["DOTENV_KEY"] = 'dotenv://:key_ddcaa26504cd70a6fef9801901c3981538563a1767c297cb8416e8a38c62fe00@dotenv.org/vault/.env.vault?environment=development';

        $dotenvVault = DotenvVault::createUnsafeMutable([__DIR__, self::$folder]);
        $dotenvVault->load();

        self::assertEquals(getenv('ALPHA'), 'zeta');
    }
}


tests/DotenvVault/Decrypter/DecrypterTest.php
<?php

declare(strict_types=1);

namespace DotenvVault\Tests\Decrypter;

use Exception;
use DotenvVault\Decrypter\Decrypter;
use DotenvVault\Decrypter\DecrypterInterface;
use PHPUnit\Framework\TestCase;

final class DecrypterTest extends TestCase
{
    protected $encrypted;

    protected function setUp(): void {
        $this->encrypted = 's7NYXa809k/bVSPwIAmJhPJmEGTtU0hG58hOZy7I0ix6y5HP8LsHBsZCYC/gw5DDFy5DgOcyd18R';
    }

    public function testDecrypterInstanceOf()
    {
        self::assertInstanceOf(DecrypterInterface::class, new Decrypter());
    }

    public function testFullDecrypt()
    {
        $keyStr = 'ddcaa26504cd70a6fef9801901c3981538563a1767c297cb8416e8a38c62fe00';

        $result = (new Decrypter())->decrypt($this->encrypted, $keyStr);

        self::assertEquals($result, "# development@v6\nALPHA=\"zeta\"");
    }

    public function testDecryptWhenKeyIsTooShort()
    {
        $keyStr = 'vlt_tooshort';

        $this->expectExceptionMessage('INVALID_DOTENV_KEY: It must be 64 characters long (or more)');

        $result = (new Decrypter())->decrypt($this->encrypted, $keyStr);

        self::assertEquals($result, "# development@v6\nALPHA=\"zeta\"");
    }

    public function testDecryptWhenKeyIsWrong()
    {
        $keyStr = 'AAcaa26504cd70a6fef9801901c3981538563a1767c297cb8416e8a38c62fe00';

        $this->expectExceptionMessage('DECRYPTION_FAILED: Please check your DOTENV_KEY');

        $result = (new Decrypter())->decrypt($this->encrypted, $keyStr);

        self::assertEquals($result, "# development@v6\nALPHA=\"zeta\"");
    }

    public function testDecryptEmptyEncryptedText()
    {
        $keyStr = 'ddcaa26504cd70a6fef9801901c3981538563a1767c297cb8416e8a38c62fe00';

        $this->expectExceptionMessage('MISSING_CIPHERTEXT: It must be a non-empty string');

        $result = (new Decrypter())->decrypt('', $keyStr);
    }
}


src/DotEnvVault.php
<?php

namespace DotenvVault;

use Dotenv\Dotenv;
use Dotenv\Loader\Loader;
use Dotenv\Loader\LoaderInterface;
use Dotenv\Parser\Parser;
use Dotenv\Parser\ParserInterface;
use Dotenv\Repository\RepositoryBuilder;
use Dotenv\Repository\RepositoryInterface;
use Dotenv\Store\StoreBuilder;
use Dotenv\Store\StoreInterface;
use Exception;

class DotEnvVault extends Dotenv
{
    /** @var StoreInterface */
    private $store;
    /** @var ParserInterface */
    private $parser;
    /** @var LoaderInterface */
    private $loader;
    /** @var RepositoryInterface */
    private $repository;
    /** @var string|null */
    private $dotenv_key;

    public function __construct(
        StoreInterface $store,
        ParserInterface $parser,
        LoaderInterface $loader,
        RepositoryInterface $repository
    ) {
        $this->store = $store;
        $this->parser = $parser;
        $this->loader = $loader;
        $this->repository = $repository;
    }

    /**
     * @throws DotEnvVaultError
     */
    public function load()
    {
        $vaultEntries = $this->parser->parse($this->store->read());
        $vaultArr = $this->loader->load($this->repository, $vaultEntries);

        $this->dotenv_key = $this->repository->get("DOTENV_KEY");
        if (empty($this->dotenv_key)) {
            return $vaultArr;
        }

        $decryptedStr = $this->parse_vault();

        // parsing plaintext and loading into repository
        $decryptedEntries = $this->parser->parse($decryptedStr);
        return $this->loader->load($this->repository, $decryptedEntries);
    }

    /**
     * Parses encrypted values within vault.
     * @throws DotEnvVaultError
     */
    public function parse_vault(): ?string
    {
        $dotenv_keys = explode(',', $this->dotenv_key);
        $keys = [];
        foreach($dotenv_keys as $key)
        {
            // parse DOTENV_KEY, format is a URI.
            $uri = parse_url(trim($key));

            // get encrypted key
            $pass = $uri['pass'];

            // Get environment from query params.
            parse_str($uri['query'], $params);
            if (!($vault_environment = $params['environment'] ?? false)) {
                throw new DotEnvVaultError('INVALID_DOTENV_KEY: Missing environment part.');
            }

            # Getting ciphertext from correct environment in .env.vault
            $vault_environment = strtoupper($vault_environment);
            $environment_key = "DOTENV_VAULT_{$vault_environment}";

            if (!($ciphertext = $this->repository->get($environment_key))) {
                throw new DotEnvVaultError("NOT_FOUND_DOTENV_ENVIRONMENT: Cannot locate environment {$environment_key} in your .env.vault file. Run 'npx dotenv-vault build' to include it.");
            }

            $keys[] = ['encrypted_key' => $pass, 'ciphertext' => $ciphertext];
        }
        return $this->key_rotation($keys);
    }

    /**
     * @throws DotEnvVaultError
     */
    private function key_rotation($keys): ?string
    {
        $count = count($keys);
        foreach($keys as $index=>$value) {
            $decrypt = $this->decrypt($value['ciphertext'], $value['encrypted_key']);

            if ($decrypt === false && $index + 1 >= $count){
                throw new DotEnvVaultError('INVALID_DOTENV_KEY: Key must be valid.');
            }
            elseif($decrypt === false){
                continue;
            }
            else{
                return $decrypt;
            }
        }

        return null;
    }

    private function decrypt($data, $secret)
    {
        $secret = hex2bin(substr($secret, 4, strlen($secret)));
        $data = base64_decode($data, true);
        $nonce = substr($data, 0, 12);
        $tag = substr($data, -16);
        $ciphertext = substr($data, 12, -16);

        try {
            return openssl_decrypt(
                $ciphertext,
                'aes-256-gcm',
                $secret,
                OPENSSL_RAW_DATA,
                $nonce,
                $tag
            );
        } catch (Exception $e) {
            return false;
        }
    }

    public static function createImmutable($paths, $names = null, bool $shortCircuit = true, string $fileEncoding = null)
    {
        $repository = RepositoryBuilder::createWithDefaultAdapters()->immutable()->make();

        return self::create($repository, $paths, $names, $shortCircuit, $fileEncoding);
    }

    public static function create(RepositoryInterface $repository, $paths, $names = null, bool $shortCircuit = true, string $fileEncoding = null)
    {
        $builder = $names === null ? StoreBuilder::createWithDefaultName() : StoreBuilder::createWithNoNames();

        foreach ((array) $paths as $path) {
            $builder = $builder->addPath($path);
        }

        foreach ((array) $names as $name) {
            $builder = $builder->addName($name);
        }

        if ($shortCircuit) {
            $builder = $builder->shortCircuit();
        }

        return new self($builder->fileEncoding($fileEncoding)->make(), new Parser(), new Loader(), $repository);
    }
}


src/Decrypter/Decrypter.php
<?php

declare(strict_types=1);

namespace DotenvVault\Decrypter;

use Exception;

final class Decrypter implements DecrypterInterface
{
    /**
     * Decrypt encrypted content into a string
     *
     * @param string $content
     * @param string $keyStr
     *
     * @throws \Exception
     *
     * @return string
     */
    public function decrypt(string $encrypted, string $keyStr)
    {
        if ($encrypted === null || !is_string($encrypted) || strlen($encrypted) < 1) {
          $msg = 'MISSING_CIPHERTEXT: It must be a non-empty string';
          throw new Exception($msg);
        }

        // grab last 64 to permit keys like vlt_64 or custom_64
        $last64 = substr($keyStr, -64);

        // must be 64 characters long
        if (strlen($last64) !== 64) {
          $msg = 'INVALID_DOTENV_KEY: It must be 64 characters long (or more)';
          throw new Exception($msg);
        }

        // check key length is good INVALID_DOTENV_KEY: It must be 64 characters long (or more)
        $key = hex2bin($last64);

        // base64 decode
        $decoded = base64_decode($encrypted, true);

        // determine cipher and pull out nonce and tag
        $ciphertext = substr($decoded, 12, -16);
        $nonce = substr($decoded, 0, 12);
        $tag = substr($decoded, -16);
    
        try {
          $plaintext = openssl_decrypt($ciphertext, 'aes-256-gcm', $key, OPENSSL_RAW_DATA, $nonce, $tag);

          if ($plaintext === false) {
            $msg = 'DECRYPTION_FAILED: Please check your DOTENV_KEY';
            throw new Exception($msg);
          } else {
            return $plaintext;
          }
        } catch (ExceptionType $e) {
          $msg = 'DECRYPTION_FAILED: Please check your DOTENV_KEY';
          throw new Exception($msg);
        }
    }
}


src/DotenvVaultBackup.php
<?php

namespace DotenvVault;
use Dotenv\Dotenv;
use Dotenv\Loader\Loader;
use Dotenv\Loader\LoaderInterface;
use Dotenv\Parser\Parser;
use Dotenv\Parser\ParserInterface;
use Dotenv\Repository\RepositoryBuilder;
use Dotenv\Repository\RepositoryInterface;
use Dotenv\Store\StoreBuilder;
use Dotenv\Store\StoreInterface;
use Exception;

class DotEnvVaultError extends Exception { }

class DotEnvVault extends Dotenv
{
    private $store;
    private $parser;
    private $loader;
    private $repository;
    private $dotenv_key;
    public function __construct(
        StoreInterface $store,
        ParserInterface $parser,
        LoaderInterface $loader,
        RepositoryInterface $repository
    )
    {
        $this->store = $store;
        $this->parser = $parser;
        $this->loader = $loader;
        $this->repository = $repository;
    }

    public function load()
    {
        $this->dotenv_key = getenv("DOTENV_KEY");
        if ($this->dotenv_key !== false){

            $entries = $this->parser->parse($this->store->read());
            $this->loader->load($this->repository, $entries);

            $plaintext = $this->parse_vault();

            // parsing plaintext and loading to getenv 
            $vault_entries = $this->parser->parse($plaintext);
            return $this->loader->load($this->repository, $vault_entries);
        }
        else {
            $entries = $this->parser->parse($this->store->read());

            return $this->loader->load($this->repository, $entries);
        }
    }

    public function parse_vault()
    {
        $dotenv_keys = explode(',', $this->dotenv_key);
        $keys = array();
        foreach($dotenv_keys as $key)
        {
            // parse DOTENV_KEY, format is a URI.
            $uri = parse_url(trim($key));

            // get encrypted key
            $pass = $uri['pass'];
            
            // Get environment from query params.
            parse_str($uri['query'], $params);
            $vault_environment = $params['environment'] or throw new DotEnvVaultError('INVALID_DOTENV_KEY: Missing environment part.');

            # Getting ciphertext from correct environment in .env.vault
            $vault_environment = strtoupper($vault_environment);
            $environment_key = "DOTENV_VAULT_{$vault_environment}";

            $ciphertext = getenv("{$environment_key}") or throw new DotEnvVaultError("NOT_FOUND_DOTENV_ENVIRONMENT: Cannot locate environment {$environment_key} in your .env.vault file. Run 'npx dotenv-vault build' to include it.");
            
            array_push($keys, array('encrypted_key' => $pass, 'ciphertext' => $ciphertext));
        }
        return $this->key_rotation($keys);
    }

    private function key_rotation($keys){
        $count = count($keys);
        foreach($keys as $index=>$value) {
            $decrypt = $this->decrypt($value['ciphertext'], $value['encrypted_key']);

            if ($decrypt == false && $index + 1 >= $count){
                throw new DotEnvVaultError('INVALID_DOTENV_KEY: Key must be valid.');
            }
            elseif($decrypt == false){
                continue;
            }
            else{
                return $decrypt;
            }
        }
    }

    public static function decrypt($encrypted, $keyStr)
    {
        // grab last 64 to permit keys like vlt_64 or custom_64
        $last64 = substr($keyStr, -64);

        if (strlen($last64) !== 64) {
          $msg = 'INVALID_DOTENV_KEY: It must be 64 characters long (or more)';
          throw new Exception($msg);
        }

        // check key length is good INVALID_DOTENV_KEY: It must be 64 characters long (or more)
        $key = hex2bin($last64);

        // base64 decode
        $decoded = base64_decode($encrypted, true);

        // determine cipher and pull out nonce and tag
        $ciphertext = substr($decoded, 12, -16);
        $nonce = substr($decoded, 0, 12);
        $tag = substr($decoded, -16);
    
        try {
          $plaintext = openssl_decrypt($ciphertext, 'aes-256-gcm', $key, OPENSSL_RAW_DATA, $nonce, $tag);

          if ($plaintext === false) {
            $msg = 'DECRYPTION_FAILED: Please check your DOTENV_KEY';
            throw new Exception($msg);
          } else {
            return $plaintext;
          }
        } catch (ExceptionType $e) {
          $msg = 'DECRYPTION_FAILED: Please check your DOTENV_KEY';
          throw new Exception($msg);
        }
    }

    /**
     * Create a new immutable dotenvVault instance with default repository.
     *
     * @param string|string[]      $paths
     * @param string|string[]|null $names
     * @param bool                 $shortCircuit
     * @param string|null          $fileEncoding
     *
     * @return \DotenvVault\DotenvVault
     */
    public static function createImmutable($paths, $names = null, bool $shortCircuit = true, string $fileEncoding = null)
    {  
        $repository = RepositoryBuilder::createWithDefaultAdapters()->immutable()->make();

        return self::create($repository, $paths, $names, $shortCircuit, $fileEncoding);
    }

    /**
     * Create a new immutable dotenvVault instance with default repository with the putenv adapter.
     *
     * @param string|string[]      $paths
     * @param string|string[]|null $names
     * @param bool                 $shortCircuit
     * @param string|null          $fileEncoding
     *
     * @return \DotenvVault\DotenvVault
     */
    public static function createUnsafeImmutable($paths, $names = null, bool $shortCircuit = true, string $fileEncoding = null)
    {
        $repository = RepositoryBuilder::createWithDefaultAdapters()
            ->addAdapter(PutenvAdapter::class)
            ->immutable()
            ->make();

        return self::create($repository, $paths, $names, $shortCircuit, $fileEncoding);
    }

    public static function create(RepositoryInterface $repository, $paths, $names = null, bool $shortCircuit = true, string $fileEncoding = null)
    {
        $builder = $names === null ? StoreBuilder::createWithDefaultName() : StoreBuilder::createWithNoNames();

        foreach ((array) $paths as $path) {
            $builder = $builder->addPath($path);
        }

        foreach ((array) $names as $name) {
            $builder = $builder->addName($name);
        }

        if ($shortCircuit) {
            $builder = $builder->shortCircuit();
        }

        return new self($builder->fileEncoding($fileEncoding)->make(), new Parser(), new Loader(), $repository);
    }
}
